# How do I select a pandas Series from a DataFrame

Tuto on pandas by Data School - Exercice performed by Dorian.H Mekni | Wed 12 Feb 2020

In [1]:
import pandas as pd


In [4]:
df_new = pd.read_table('https://bit.ly/uforeports', sep=',')

In [5]:
df_new

City Colors Reported Shape Reported State  \
0                    Ithaca             NaN       TRIANGLE    NY   
1               Willingboro             NaN          OTHER    NJ   
2                   Holyoke             NaN           OVAL    CO   
3                   Abilene             NaN           DISK    KS   
4      New York Worlds Fair             NaN          LIGHT    NY   
...                     ...             ...            ...   ...   
18236            Grant Park             NaN       TRIANGLE    IL   
18237           Spirit Lake             NaN           DISK    IA   
18238           Eagle River             NaN            NaN    WI   
18239           Eagle River             RED          LIGHT    WI   
18240                  Ybor             NaN           OVAL    FL   

                   Time  
0        6/1/1930 22:00  
1       6/30/1930 20:00  
2       2/15/1931 14:00  
3        6/1/1931 13:00  
4       4/18/1933 19:00  
...                 ...  
18236  12/31/2000 23:00  
18237  12/31/2000 23:00  
18238  12/31/2000 23:45  
18239  12/31/2000 23:45  
18240  12/31/2000 23:59  

[18241 rows x 5 columns]

We are notified above that our dataframe is made of 18241 rows and 5 columns

In [21]:
df_new = pd.read_csv('https://bit.ly/uforeports')

In [22]:
df_new

City Colors Reported Shape Reported State  \
0                    Ithaca             NaN       TRIANGLE    NY   
1               Willingboro             NaN          OTHER    NJ   
2                   Holyoke             NaN           OVAL    CO   
3                   Abilene             NaN           DISK    KS   
4      New York Worlds Fair             NaN          LIGHT    NY   
...                     ...             ...            ...   ...   
18236            Grant Park             NaN       TRIANGLE    IL   
18237           Spirit Lake             NaN           DISK    IA   
18238           Eagle River             NaN            NaN    WI   
18239           Eagle River             RED          LIGHT    WI   
18240                  Ybor             NaN           OVAL    FL   

                   Time  
0        6/1/1930 22:00  
1       6/30/1930 20:00  
2       2/15/1931 14:00  
3        6/1/1931 13:00  
4       4/18/1933 19:00  
...                 ...  
18236  12/31/2000 23:00  
18237  12/31/2000 23:00  
18238  12/31/2000 23:45  
18239  12/31/2000 23:45  
18240  12/31/2000 23:59  

[18241 rows x 5 columns]

In [6]:
type(df_new)
#This indicates the datatype.

pandas.core.frame.DataFrame

In [7]:
df_new.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

In [8]:
df_new['City']
#Bracket notation is another option to select specific pandas series

0                      Ithaca
1                 Willingboro
2                     Holyoke
3                     Abilene
4        New York Worlds Fair
                 ...         
18236              Grant Park
18237             Spirit Lake
18238             Eagle River
18239             Eagle River
18240                    Ybor
Name: City, Length: 18241, dtype: object

In [9]:
type(df_new['City'])
# By this operation we have the confirmation that it is a series.

pandas.core.series.Series

In [10]:
df_new.City
#Here is a short cut known as a dot notation '.'

0                      Ithaca
1                 Willingboro
2                     Holyoke
3                     Abilene
4        New York Worlds Fair
                 ...         
18236              Grant Park
18237             Spirit Lake
18238             Eagle River
18239             Eagle River
18240                    Ybor
Name: City, Length: 18241, dtype: object

Everytime a series is added to a daframe, all its names become automatically an attribute 
of that dataframe. 

In [11]:
df_new.Colors_Reported

AttributeError: 'DataFrame' object has no attribute 'Colors_Reported'

That doesn't work ! let's try something else...

In [12]:
df_new.ColorsReported

AttributeError: 'DataFrame' object has no attribute 'ColorsReported'

That doesn't work either! 

In [13]:
df_new['Colors Reported']
#This is bracket notation in use and it works. 

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
18236    NaN
18237    NaN
18238    NaN
18239    RED
18240    NaN
Name: Colors Reported, Length: 18241, dtype: object

# Bonus tips: How to I create a new pandas series within a dataframe ? 

In [14]:
'ab' + 'cd' 

'abcd'

In [15]:
df_new.City + df_new.State


0                      IthacaNY
1                 WillingboroNJ
2                     HolyokeCO
3                     AbileneKS
4        New York Worlds FairNY
                  ...          
18236              Grant ParkIL
18237             Spirit LakeIA
18238             Eagle RiverWI
18239             Eagle RiverWI
18240                    YborFL
Length: 18241, dtype: object

In [17]:
df_new.City + ', ' + df_new.State
#We insert a ', ' to get the bellow outcome. 

0                      Ithaca, NY
1                 Willingboro, NJ
2                     Holyoke, CO
3                     Abilene, KS
4        New York Worlds Fair, NY
                   ...           
18236              Grant Park, IL
18237             Spirit Lake, IA
18238             Eagle River, WI
18239             Eagle River, WI
18240                    Ybor, FL
Length: 18241, dtype: object

In [18]:
df_new.Location = df_new.City + ', ' + df_new.State

<ipython-input-18-5b6a8c644eb7>:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df_new.Location = df_new.City + ', ' + df_new.State


In [19]:
df_new['Location'] = df_new.City + ', ' + df_new.State

In [20]:
df_new.head()
# We have now integrated successfully our new series 'Location' within our dataframe.

City Colors Reported Shape Reported State             Time  \
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00   
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00   
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00   
3               Abilene             NaN           DISK    KS   6/1/1931 13:00   
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00   

                   Location  
0                Ithaca, NY  
1           Willingboro, NJ  
2               Holyoke, CO  
3               Abilene, KS  
4  New York Worlds Fair, NY